### PR TITLE
Add timing for block availability

### PR DESCRIPTION
**Pull request by michaelsproul**
## Issue Addressed

Presently the block import metrics that measure block observation delay and import delay (time to import a block) are unaware of blob-processing and related delays.

This can lead to the `import_delay` seeming artificially high because it also factors in the time spent waiting for blobs to arrive.

## Proposed Changes

- [x] Add an `available` timestamp to the block delay cache which records when a block became fully available.
- [x] Add a similar `attestable` timestamp that records when a block became capable of being attested to (this allows us to ignore the non-critical disk writes and similar that are present in the `import_delay`). Revived from an old PR: https://github.com/sigp/lighthouse/pull/3232.
- [x] Log these delays _every slot_ (not just for late blocks), in milliseconds.
- [x] Add a `debug` log with the exact time taken by each `newPayload` call to the EL. This helps per-block investigations and is clearer and more accurate than the metric histogram. We can also use the exact `ms` value combined with log timestamp and the `observed_delay` to calculate the time taken doing other stuff prior to calling the EL (`log_timestamp - slot_start - time_taken_ms - observed_delay`).
- [x] Add matching prometheus metrics.

## Additional Info

The `available` delay currently reported by this PR is much higher than I expected, and sometimes seems to include things like snapshot cache misses. I need to do some more investigating to make sure I've put the observation of availability at the earliest possible point.

### CODE REVIEW

File: beacon_node/beacon_chain/src/data_availability_checker.rs [hunk 1]
```diff
@@ -318,10 +332,15 @@ impl<T: BeaconChainTypes> DataAvailabilityChecker<T> {
                     if self.blobs_required_for_block(&block) {
                         results.push(MaybeAvailableBlock::AvailabilityPending { block_root, block })
                     } else {
+                        let available_timestamp = self
+                            .slot_clock
+                            .now_duration()
+                            .ok_or(AvailabilityCheckError::SlotClockError)?;
                         results.push(MaybeAvailableBlock::Available(AvailableBlock {
```
Comment (ethDreamer) [inline beacon_node/beacon_chain/src/data_availability_checker.rs]: I'm fairly certain this function is only called during range-sync or backfill. So these blocks are not related to the current slot. Also due to the way sync works, none of the blocks hit this cache until all the data for all the blocks in the request are downloaded. Thus it's a bit meaningless to measure how far into the current slot they became available. Might be a good idea to exclude them from the data?

File: beacon_node/beacon_chain/src/data_availability_checker.rs [hunk 2]
```diff
@@ -543,6 +567,8 @@ pub struct AvailableBlock<E: EthSpec> {
     block_root: Hash256,
     block: Arc<SignedBeaconBlock<E>>,
     blobs: Option<BlobSidecarList<E>>,
+    /// Timestamp at which this block first became available (UNIX timestamp, time since 1970).
+    available_timestamp: Duration,
 }
```
Comment (ethDreamer) [inline beacon_node/beacon_chain/src/data_availability_checker.rs:478]: Storing the timestamp here will force you to include the timestamp for backfill & range-sync blocks which will only pollute the data. I think a better way would be to include this timestamp in the [`AvailableExecutedBlock`](https://github.com/sigp/lighthouse/blob/v5.1.3/beacon_node/beacon_chain/src/block_verification_types.rs#L230) & then populate it inside [`make_available()`](https://github.com/sigp/lighthouse/blob/v5.1.3/beacon_node/beacon_chain/src/data_availability_checker/overflow_lru_cache.rs#L117). This will probe what I believe you actually want as `make_available()` is the final destination of:
 - [`put_rpc_blobs()`](https://github.com/sigp/lighthouse/blob/v5.1.3/beacon_node/beacon_chain/src/data_availability_checker.rs#L203)
 - [`put_gossip_blob()`](https://github.com/sigp/lighthouse/blob/v5.1.3/beacon_node/beacon_chain/src/data_availability_checker.rs#L224)
 - [`put_pending_executed_block()`](https://github.com/sigp/lighthouse/blob/v5.1.3/beacon_node/beacon_chain/src/data_availability_checker.rs#L234)
 
 which are the 3 entry points for which new data can come in & complete a block.
 
 The other entry points:
 - [`verify_kzg_for_rpc_block()`](https://github.com/sigp/lighthouse/blob/v5.1.3/beacon_node/beacon_chain/src/data_availability_checker.rs#L247)
 - [`verify_kzg_for_rpc_blocks()`](https://github.com/sigp/lighthouse/blob/v5.1.3/beacon_node/beacon_chain/src/data_availability_checker.rs#L291)
 
 Don't actually hit the DA cache as they assume all blobs are already present.

File: beacon_node/beacon_chain/src/data_availability_checker.rs (hunk 1 above)
Comment (michaelsproul) [reply beacon_node/beacon_chain/src/data_availability_checker.rs]: got it!

File: beacon_node/beacon_chain/src/beacon_chain.rs [hunk 3]
```diff
@@ -2998,22 +3007,21 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
...
         metrics::inc_counter(&metrics::BLOCK_PROCESSING_REQUESTS);
 
+        let block_slot = unverified_block.block().slot();
+
         // Set observed time if not already set. Usually this should be set by gossip or RPC,
         // but just in case we set it again here (useful for tests).
-        if let (Some(seen_timestamp), Some(current_slot)) =
-            (self.slot_clock.now_duration(), self.slot_clock.now())
-        {
+        #[cfg(test)]
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/beacon_chain.rs]: I don't see any point in adding this line. It's fine for us to do this outside of tests, _and_ this line will only affect tests for the `beacon_chain` crate itself. Other crates that compile `beacon_chain` and use the test harness cannot supply `--cfg test`, so they won't turn this off.

File: beacon_node/beacon_chain/src/beacon_chain.rs [hunk 4]
```diff
@@ -3256,6 +3269,16 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
             consensus_context,
         } = import_data;
 
+        // This is the time since start of the slot where all the components of the block have become available
+        let delay = get_slot_delay_ms(timestamp_now(), block.slot(), &self.slot_clock);
+        metrics::set_gauge(&metrics::BLOCK_AVAILABILITY_DELAY, delay.as_millis() as i64);
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/beacon_chain.rs]: should probably use `block.available_timestamp()` here instead

File: beacon_node/beacon_chain/src/data_availability_checker.rs (hunk 2 above)
Comment (michaelsproul) [reply beacon_node/beacon_chain/src/data_availability_checker.rs:491]: I tried doing this just now but got stuck because to fill in the `available_timestamp` on `AvailableExecutedBlock` here would require it to be part of the `AvailableBlock` anyway:

https://github.com/sigp/lighthouse/blob/3058b96f2560f1da04ada4f9d8ba8e5651794ff6/beacon_node/beacon_chain/src/block_verification_types.rs#L194-L200

File: beacon_node/beacon_chain/src/beacon_chain.rs (hunk 3 above)
Comment (michaelsproul) [reply beacon_node/beacon_chain/src/beacon_chain.rs]: I pushed this commit to remove it d21a7f15aefaf0f73ab76d9e52183f768360df17

File: beacon_node/beacon_chain/src/beacon_chain.rs (hunk 4 above)
Comment (michaelsproul) [reply beacon_node/beacon_chain/src/beacon_chain.rs]: changed in d21a7f15aefaf0f73ab76d9e52183f768360df17

File: beacon_node/beacon_chain/src/data_availability_checker.rs (hunk 2 above)
Comment (michaelsproul) [reply beacon_node/beacon_chain/src/data_availability_checker.rs:478]: I tried doing this just now but got stuck because to fill in the `available_timestamp` on `AvailableExecutedBlock` here would require it to be part of the `AvailableBlock` anyway:

https://github.com/sigp/lighthouse/blob/3058b96f2560f1da04ada4f9d8ba8e5651794ff6/beacon_node/beacon_chain/src/block_verification_types.rs#L194-L200

**Comment by michaelsproul** [comment]
@AgeManning I pushed a commit here 73cb98281da07440795d6659ef291be8d4d47e9a which I think fixes the issues with RPC blobs not being recorded. We now record a seen timestamp for every blob, and when we call `make_available` we take the max of the seen timestamps to calculate the `blobs_available_timestamp` (this is similar to what @ethDreamer suggested here: https://github.com/sigp/lighthouse/pull/5510#discussion_r1548215254).

From watching the logs it seems to produce similar results. If you could try your metrics dash with the latest commit and confirm that it looks as expected, that would be great :pray:

**Review by AgeManning — APPROVED** [review]

**Comment by michaelsproul** [comment]
@mergifyio queue

**Comment by michaelsproul** [comment]
Marking as a breaking change because this PR deletes and changes some metrics related to block processing.

**Comment by michaelsproul** [comment]
@mergifyio queue

**Comment by michaelsproul** [comment]
@mergifyio requeue

File: beacon_node/beacon_chain/src/metrics.rs [hunk 5]
```diff
@@ -847,37 +847,55 @@ lazy_static! {
...
+        "Duration between the time the block became available and the time when it was imported.",
+    );
+
+    pub static ref BEACON_BLOCK_DELAY_HEAD_IMPORTED_TIME: Result<IntGauge> = try_create_int_gauge(
+        "beacon_block_delay_head_imported_time",
+        "Duration between the time that block was imported and the time when it was set as head.",
+    );
+    pub static ref BEACON_BLOCK_DELAY_HEAD_SLOT_START_EXCEEDED_TOTAL: Result<IntCounter> = try_create_int_counter(
+        "beacon_block_delay_head_slot_start_exceeded_total",
+        "A counter that is triggered when the duration between the start of the block's slot and the current time \
```
Comment (dapplion) [inline beacon_node/beacon_chain/src/metrics.rs:898]: @michaelsproul why did you change from histogram to gauge here?

Comment (michaelsproul) [reply beacon_node/beacon_chain/src/metrics.rs:898]: that was @AgeManning 

thinking was to get exact values rather than buckets

Comment (dapplion) [reply beacon_node/beacon_chain/src/metrics.rs:898]: hmm I guess the average observation frequency is 1 every 12 seconds so kinda ok? Still weird tho
